tools/chpldef/Message.h

```diff
@@ -0,0 +1,396 @@
+  Error error_ = Message::OK; 
+  Status status_ = Message::PENDING;
+  std::string note_;
+
+protected:
+  inline void markProgressing() { status_ = Message::PROGRESSING; }
+  inline void markCompleted() { status_ = Message::COMPLETED; }
+  inline void markFailed(Error error, std::string note=std::string()) {
+    status_ = Message::FAILED;  
+    error_ = error;
+    note_ = std::move(note);
+  }
+
+  static bool isIdValid(const JsonValue& id);
+
+  Message(Message::Tag tag, JsonValue&& id, Error error,
+          std::string note)
+      : tag_(tag),
+        id_(std::move(id)),
+        error_(error),
+        note_(std::move(note)) {
+    CHPL_ASSERT(isIdValid(id_));
+  }
+
+public:
+  virtual ~Message() = default;
+
+  /** Create a request given a JSON value. */
+  static chpl::owned<Message> request(Server* ctx, JsonValue&& json);
+
+  /** Create a response to a handled message. The status of the message 
+      must be COMPLETED or FAILED, or else nothing is returned. */
+  static opt<Response> response(Server* ctx, const Message* msg);
+
+  /** The tag for this message. */
+  inline Tag tag() const { return tag_; }
+
+  /** Returns 'true' if this message has an error value. */
+  inline bool hasError() const { return error_ != Message::OK; }
+
+  /** If this message failed, return the reason why. */
+  inline Message::Error error() const { return error_; }
+
+  /** If this message failed, return a possible note explaining why. */
+  const std::string& note() const { return note_; }
+
+  /** Get the numeric value of an error code. */
+  static inline int64_t errorToInt(Error e) {
+    return static_cast<int64_t>(e);
+  }
+
+  /** Get the numeric value of an error code. */
+  inline int64_t errorToInt() const { return errorToInt(error_); }
+
+  /** Return the current status of this message. */
+  inline Message::Status status() const { return status_; }
+
+  /** Returns 'true' if this message is marked 'COMPLETED'. */
+  inline bool isCompleted() const { return status_ == Message::COMPLETED; }
+
+  /** If 'true', this message is a response. */
+  inline bool isResponse() const { return tag_ == Message::RESPONSE; }
+
+  /** If 'true', the server is sending this message to the client. */
+  bool isOutbound() const;
+
+  /** If 'true', then this message does not need a response. */
+  bool isNotification() const;
+
+  /** Returns the expected JSON-RPC method name for this message. */
+  const char* jsonRpcMethodName() const;
+
+  /** Returns the message ID. */
+  inline const JsonValue& id() const { return id_; }
+
+  /** Returns the message ID as a string. */
+  std::string idToString() const;
+
+  /** Convert a given message tag to string. */
+  static const char* tagToJsonRpcMethodName(Tag tag);
+
+  /** Convert from JSON-RPC method name to message tag. */
+  static Tag jsonRpcMethodNameToTag(const char* str);
+
+  /** Convert from JSON-RPC method name to message tag. */
+  static Tag jsonRpcMethodNameToTag(std::string str);
+
```

Comment:
Could be combined into one using an LLVM StringRef perhaps?

Reply:
Could also just have the single `std::string` overload I guess.